src/Microsoft.Framework.Configuration/ConfigurationSection.cs

```diff
@@ -38,64 +41,79 @@ public IEnumerable<IConfigurationSource> Sources
+        {
+            get
+            {
+                if (_value != null)
+                {
+                    return _value;
+                }
+                else
+                {
+                    _value = Get(_key);
+                    return _value;
+                }
+            }
+
+            set
+            {
+                Set(_key, value);
+            }
         }
 
-        public bool TryGet([NotNull] string key, out string value)
+        public IConfigurationSection GetSection(string key)
+        {
+            return new ConfigurationFocus(this, key + Constants.KeyDelimiter);
+        }
+
+        public IEnumerable<IConfigurationSection> GetChildren()
+        {
+            return GetChildrenImplementation(_key + Constants.KeyDelimiter);
+        }
+
+        private string Get([NotNull] string key)
         {
             // If a key in the newly added configuration source is identical to a key in a 
             // formerly added configuration source, the new one overrides the former one.
             // So we search in reverse order, starting with latest configuration source.
             foreach (var src in _sources.Reverse())
             {
+                string value = null;
                 if (src.TryGet(key, out value))
                 {
-                    return true;
+                    return value;
                 }
             }
-            value = null;
-            return false;
+
+            return null;
         }
 
-        public void Set([NotNull] string key, [NotNull] string value)
+        private void Set([NotNull] string key, [NotNull] string value)
         {
             if (!_sources.Any())
             {
                 throw new InvalidOperationException(Resources.Error_NoSources);
             }
-            foreach (var src in _sources)
-            {
-                src.Set(key, value);
-            }
-        }
 
-        public void Reload()
-        {
+            _key = key;
+            _value = value;
+
             foreach (var src in _sources)
             {
-                src.Load();
+                src.Set(key, value);
             }
         }
 
-        public IConfiguration GetConfigurationSection(string key)
-        {
-            return new ConfigurationFocus(this, key + Constants.KeyDelimiter);
-        }
-
-        public IEnumerable<KeyValuePair<string, IConfiguration>> GetConfigurationSections()
-        {
-            return GetConfigurationSectionsImplementation(string.Empty);
-        }
-
-        public IEnumerable<KeyValuePair<string, IConfiguration>> GetConfigurationSections([NotNull] string key)
-        {
-            return GetConfigurationSectionsImplementation(key + Constants.KeyDelimiter);
-        }
-
-        private IEnumerable<KeyValuePair<string, IConfiguration>> GetConfigurationSectionsImplementation(string prefix)
+        private IEnumerable<IConfigurationSection> GetChildrenImplementation(string prefix)
```

Comment:
Move to calling method and move calling method implementation to common base class.